File: Functions/camera_calibration.py
#!/usr/bin/env python3	
# -*- coding: utf-8 -*-

import numpy as np
from matplotlib import pyplot as plt
import cv2
import glob
import time
import pickle
import sys


# Termination criteria
criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 100, 0.001)

# Prepare object points, like (0,0,0), (1,0,0), (2,0,0) ....,(6,5,0)
objp = np.zeros((9*6,3), np.float32)
objp[:,:2] = np.mgrid[0:9,0:6].T.reshape(-1,2)

# Arrays to store object points and image points from all the images.
objpoints = [] # 3d point in real world space
imgpoints = [] # 2d points in image plane.
size_pattern = 22.5 # mm


cap = cv2.VideoCapture(0)
#cap.set(cv2.CAP_PROP_BUFFERSIZE,1)
cap.set(cv2.CAP_PROP_FRAME_WIDTH,600);
cap.set(cv2.CAP_PROP_FRAME_HEIGHT,400);

while(True):
    
    # Capture frame-by-frame ignoring camera buffer
    for i in range(5):
        ret, frame = cap.read()
    
    # Change color and print
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    
    # Find the chess board corners
    ret, corners = cv2.findChessboardCorners(gray, (9,6),None)
    print(ret)
    # If found, add object points, image points (after refining them)
    if ret == True:
        objpoints.append(objp*size_pattern)
        print(len(objpoints))
        corners2 = cv2.cornerSubPix(gray,corners,(11,11),(-1,-1),criteria)
        imgpoints.append(corners2)

        # Draw and display the corners
        img = cv2.drawChessboardCorners(gray, (9,6), corners2,ret)
        cv2.imshow('img',img)
        cv2.waitKey(50)
        #cv2.destroyAllWindows()

    if len(objpoints)>49:
        cap.release()
        break

# Release the capture
cv2.destroyAllWindows()
cap.release()


print('calculating...')
ret, camera_matrix, dist_coef, rvecs, tvecs = cv2.calibrateCamera(objpoints, imgpoints, gray.shape[::-1],None,None)


# Save calibration data
print('saving data...')
data = {"camera_matrix":camera_matrix, "dist_coef":dist_coef}

path = sys.argv[1] + '/data.pkl')
output = open(path, 'wb')
pickle.dump(data, output)
output.close()

'''
# Load calibration data
pkl_file = open('data.pkl', 'rb')
data2 = pickle.load(pkl_file)
pkl_file.close()

camera_matrix = data2['camera_matrix']
dist_coef = data2['dist_coef']


# 2D image points.
image_points = np.array([(32, 623), (1237, 650), (1016, 229), (318, 224)], dtype="double")
 
# 3D model points.
model_points = np.array([(0, 0, 0), (297, 0, 0), (297, 210, 0), (0, 210, 0)], dtype="double")


# Find rotation and translation
(success, rotation_vector, translation_vector) = cv2.solvePnP(model_points, image_points, camera_matrix, dist_coef, cv2.SOLVEPNP_P3P)

# Estimate 2D point
(point2D, jacobian) = cv2.projectPoints(np.array([(279, 0, 0)], dtype="double"), rotation_vector, translation_vector, camera_matrix, dist_coef)


camera_matrix_inv = np.linalg.inv(camera_matrix)

rotation_matrix = cv2.Rodrigues(rotation_vector)[0]
rotation_matrix_inv = np.linalg.inv(rotation_matrix)


# Estimate 3D point
z = 0
uvPoint = np.array([1237,650,1]).reshape(3,1)

s = (z + np.matmul(rotation_matrix_inv, translation_vector)[2])/(np.matmul(np.matmul(rotation_matrix_inv, camera_matrix_inv),uvPoint)[2])
p = np.matmul(rotation_matrix_inv, (np.matmul(camera_matrix_inv, s * uvPoint) - translation_vector))
'''

File: Functions/camera_calibration.cpp
#include "opencv2/opencv.hpp"
#include <iostream>
#include <vector>
#include <unistd.h>

using namespace std;
using namespace cv;

int main(){

	int numBoards = 100;
	int board_w = 9;
	int board_h = 6;
	int success = 0;
	float squareLenght = 22.5;

	Size board_sz = Size(board_w, board_h);
	int board_n = board_w*board_h;


	vector<vector<Point3f> > calib_objectPoints;
	vector<vector<Point2f> > calib_imagePoints;
	vector<Point2f> corners;

	vector<Point3f> obj;
        for(int i = 0; i<board_h; i++){
            for(int j=0;j<board_w;j++){
                obj.push_back(cv::Point3f(float(i)*squareLenght,float(j)*squareLenght,0.0f));
            }
        }

	VideoCapture cap(-1); 
	if(!cap.isOpened()){
		cout << "Error opening video stream or file" << endl;
		return -1;
	}
	
	cap.set(CV_CAP_PROP_FRAME_WIDTH,1280);
	cap.set(CV_CAP_PROP_FRAME_HEIGHT,720);
	int frame_width = cap.get(CV_CAP_PROP_FRAME_WIDTH);
	int frame_height = cap.get(CV_CAP_PROP_FRAME_HEIGHT);


	Mat frame;
	Mat gray;

	while (success<numBoards)
	{		
	
		cap >> frame;
		cvtColor(frame, gray, CV_BGR2GRAY);
		bool patternfound = findChessboardCorners(gray,board_sz,corners);

		if (patternfound)
		{
			calib_imagePoints.push_back(corners);
			calib_objectPoints.push_back(obj);
			printf ("Corners stored\n");
			success++;

		cornerSubPix(gray, corners, Size(11, 11), Size(-1, -1), TermCriteria(CV_TERMCRIT_EPS | CV_TERMCRIT_ITER, 30, 0.1));	
		drawChessboardCorners(frame, board_sz, Mat(corners), patternfound);

		}

		imshow( "Frame", frame );
		usleep(100000);


		char c = (char)waitKey(33);
		if( c == 27 ) break;
	}
	
	cvDestroyAllWindows();
	cap.release();
	
	
	std::cout << "Stored points: " << calib_imagePoints.size() << "\n";

	Mat cameraMatrix = Mat(3, 3, CV_32FC1);
	Mat distCoeffs;
	vector<Mat> calib_rvec, calib_tvec;

	cameraMatrix.at<float>(0, 0) = 1;
	cameraMatrix.at<float>(1, 1) = 1;
	
	calibrateCamera(calib_objectPoints, calib_imagePoints, frame.size(), cameraMatrix, distCoeffs, calib_rvec, calib_tvec);


	// Read points
	std::vector<Point2f> imagePoints;
	imagePoints.push_back(Point2f(32, 623));
	imagePoints.push_back(Point2f(1237, 650));
	imagePoints.push_back(Point2f(1016, 229));
	imagePoints.push_back(Point2f(318, 224));


	std::vector<Point3f> objectPoints;
	objectPoints.push_back(Point3f(0, 0, 0));
	objectPoints.push_back(Point3f(297, 0, 0));
	objectPoints.push_back(Point3f(297, 210, 0));
	objectPoints.push_back(Point3f(0, 210, 0));


	std::cout << "There are " << imagePoints.size() << " imagePoints and " << objectPoints.size() << " objectPoints." << std::endl;


	cv::Mat rvec(3,1,cv::DataType<double>::type);
	cv::Mat tvec(3,1,cv::DataType<double>::type);

	cv::solvePnP(objectPoints, imagePoints, cameraMatrix, distCoeffs, rvec, tvec);

	std::cout << "rvec: " << rvec << std::endl;
	std::cout << "tvec: " << tvec << std::endl;

	std::vector<cv::Point2f> projectedPoints;
	cv::projectPoints(objectPoints, rvec, tvec, cameraMatrix, distCoeffs, projectedPoints);

	for(unsigned int i = 0; i < projectedPoints.size(); ++i)
	{
	std::cout << "Image point: " << imagePoints[i] << " Projected to " << projectedPoints[i] << std::endl;
	}
	
	cv::Mat rotationMatrix(3,3,cv::DataType<double>::type);
	cv::Rodrigues(rvec,rotationMatrix);

	cv::Mat uvPoint = cv::Mat::ones(3,1,cv::DataType<double>::type);
	uvPoint.at<double>(0,0) = 32.;
	uvPoint.at<double>(1,0) = 623.;
	cv::Mat tempMat, tempMat2;
	double s;
	int z = 0;
	tempMat = rotationMatrix.inv() * cameraMatrix.inv() * uvPoint;
	tempMat2 = rotationMatrix.inv() * tvec;
	s = z + tempMat2.at<double>(2,0);
	s /= tempMat.at<double>(2,0);
	std::cout << "P = " << rotationMatrix.inv() * (s * cameraMatrix.inv() * uvPoint - tvec) << std::endl;


	return 0;
}
